### PR TITLE
Add Missing Synchronization in Ack Buffer Access

### LINKED ISSUE #1029: Spring Cloud AWS SQS - some messages do not get acknowledged
**Type**: Bug

**Component**:
SQS

**Describe the bug**
JDK:17
spring-cloud-aws-dependencies : 3.1.0
org.springframework.boot: 3.2.2

When using `BatchingAcknowledgementProcessor` some messages do not get acknowledged and end up in a dead letter queue even if the processing went without any errors. Looking at the trace logs it seems that there is some race condition between purging `acksBuffer`  and adding new items to it:
`Jan 24, 2024 @ 20:09:37.170 Received message ${messageId} to ack in io.awspring.cloud.sqs.sqsListenerEndpointContainer#2-0.`
`Jan 24, 2024 @ 20:09:37.170 Removing groups [default] from buffer in io.awspring.cloud.sqs.sqsListenerEndpointContainer#2-0`

**Sample**
```
    @Bean
    fun defaultSqsListenerContainerFactory(sqsAsyncClient: SqsAsyncClient): SqsMessageListenerContainerFactory<Any> {
        return SqsMessageListenerContainerFactory.builder<Any>()
            .configure { options -> options
                .acknowledgementMode(AcknowledgementMode.MANUAL)
            }
            .sqsAsyncClient(sqsAsyncClient)
            .build()
    }

  @SqsListener(queueNames = "queue-name")
    fun process(
        sqsMessage: SqsMessage,
        acknowledgement: Acknowledgement
    ) {
        process()
       acknowledgment.acknowledge()
    }
```

### PR DESCRIPTION
**Pull request by tomazfernandes**
Fixes #1029

There was one piece of missing synchronization to the Acknowledgement Buffer that led to acknowledgements being lost.

This commit adds the missing synchronization to fix the issue.

### CODE REVIEW

**Comment by tomazfernandes** [comment]
Note: I wasn't able to reproduce acknowledgements being lost in many runs, even with a few million messages and 1000 threads, which indicates it's likely a minor racing condition.

Also, in my tests there was no noticeable difference in performance with the added synchronization, which is expected since all processing is non-blocking and the bottleneck of the overall process is I/O.

File: spring-cloud-aws-sqs/src/test/java/io/awspring/cloud/sqs/integration/BaseSqsIntegrationTest.java [hunk 1]
```diff
@@ -93,7 +95,14 @@ protected static CompletableFuture<?> createQueue(SqsAsyncClient client, String
 			if (purgeQueues) {
 				String queueUrl = v.queueUrl();
 				logger.debug("Purging queue {}", queueName);
-				return client.purgeQueue(req -> req.queueUrl(queueUrl).build());
+				return client.purgeQueue(req -> req.queueUrl(queueUrl).build())
+					.thenRun(() -> {
+						if (waitForPurge) {
+							logger.info("Waiting 30000 seconds to start sending.");
```
Comment (anna-kr) [inline spring-cloud-aws-sqs/src/test/java/io/awspring/cloud/sqs/integration/BaseSqsIntegrationTest.java:101]: 30000 seconds seems like quite long ;)

Comment (tomazfernandes) [reply spring-cloud-aws-sqs/src/test/java/io/awspring/cloud/sqs/integration/BaseSqsIntegrationTest.java:101]: Whoops... I do wish I made this mistake in the logs less often, but it just comes naturally apparently 🙂 

I've fixed this and changed a couple default test settings.

Anything else you see?

Thanks.

Comment (anna-kr) [reply spring-cloud-aws-sqs/src/test/java/io/awspring/cloud/sqs/integration/BaseSqsIntegrationTest.java:101]: it looks good to me now 🤌

**Review by anna-kr — APPROVED** [review]

**Pull request merged**